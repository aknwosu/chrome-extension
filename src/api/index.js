export const fetchPictures = async () => {
  try {
    const res = await fetch('https://jsonplaceholder.typicode.com/photos')
    return await res.clone().json()
  } catch (error) {
    return (error)
  }
}
