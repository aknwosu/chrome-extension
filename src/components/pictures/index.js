import React, { Fragment } from 'react'
import Picture from './Picture'

export const Pictures = (props) => {
  const { pictures, favoritesIDs, setFavorite } = props
  return (
    <Fragment>
      {pictures.map(picture => <Picture key={picture.id} picture={picture} favoritesIDs={favoritesIDs} setFavorite={setFavorite} />)}
    </Fragment>
  )

}

export default Pictures        