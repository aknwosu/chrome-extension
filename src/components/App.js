import React, { Component } from 'react';
import styled from 'styled-components'
import { fetchPictures, getFavoritedPics, setFavoritePic } from '../api'
import Pictures from './pictures'

export class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      pictures: [],
      isLoading: false,
      activeView: 'all',
      favoritesIDs: [],
    }
  }
  async componentDidMount() {
    this.setState({ isLoading: true })
    try {
      const pics = await fetchPictures()
      this.setState({
        pictures: pics.slice(0, 20),
        favoritesIDs: getFavoritedPics(),
        isLoading: false
      })
    } catch (error) {

    }
  }

  setFavorite = (id) => {
    setFavoritePic(id)
    this.setState({ favoritesIDs: getFavoritedPics() })
  }

  changeTab = (newTab) => {
    if (this.state.activeView !== newTab) {
      this.setState({ activeView: newTab })
    }

  }

  render() {
    let { activeView, pictures, favoritesIDs, isLoading } = this.state
    if (activeView === 'favorites') {
      pictures = pictures.filter(picture => favoritesIDs.includes(picture.id))
    }
    return (
      <Container>
        <Tab className="view-tabs">
          <TabItem isActive={activeView === 'all'} onClick={() => this.changeTab('all')}>All</TabItem>
          <TabItem isActive={activeView === 'favorites'} onClick={() => this.changeTab('favorites')}>Favorites</TabItem>
        </Tab>
        {isLoading && <div>Loading...</div>}
        {!isLoading && pictures.length > 0 && <Pictures pictures={pictures} favoritesIDs={favoritesIDs} setFavorite={this.setFavorite} />}
        {activeView === 'favorites' && pictures.length < 1 && <div>When you have favorites, they will appear here</div>}
      </Container>
    );
  }
}

export default App;

const Container = styled.div`
  text-align: center;
  background-color: #FBFBFD;
  height: 500px;
  overflow: scroll;
  margin: auto;
  width: 200px;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto";
`
const Tab = styled.div`
  display: flex;
  justify-content: space-around;
  align-items: center;
  > div {
    flex: 1
  }
  :nth-child(0) {
    border-right: 1px solid black
  }
`
export const TabItem = styled.div`
	color: ${({ isActive }) => isActive && 'green'};
`
