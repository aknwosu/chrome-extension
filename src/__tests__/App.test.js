import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme'
import { App, TabItem } from '../components/App';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { create } from 'react-test-renderer'

import { Pictures } from '../components/Pictures'

import { fetchPictures, setFavoritePic } from '../api'
configure({ adapter: new Adapter() });

jest.mock('../api');

const pictures = [
  {
    "albumId": 1,
    "id": 1,
    "title": "accusamus beatae ad facilis cum similique qui sunt",
    "url": "https://via.placeholder.com/600/92c952",
    "thumbnailUrl": "https://via.placeholder.com/150/92c952"
  },
  {
    "albumId": 1,
    "id": 2,
    "title": "reprehenderit est deserunt velit ipsam",
    "url": "https://via.placeholder.com/600/771796",
    "thumbnailUrl": "https://via.placeholder.com/150/771796"
  }
]
describe('App', () => {
  let apiMockActions;
  let wrapper
  beforeEach(() => {
    apiMockActions = {
      fetchPictures: jest.fn(),
      getFavorites: jest.fn()
    };
  });
  beforeEach(() => {
    wrapper = shallow(<App {...apiMockActions} />, { disableLifecycleMethods: true });
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('should match snapshot', () => {
    const c = create(<App />);
    expect(c.toJSON()).toMatchSnapshot();
  })
  it('calls fetchPictures on componentDidMount', async () => {
    const wrapper = shallow(<App fetchPictures={jest.fn()} />, { disableLifecycleMethods: false });
    expect(fetchPictures).toHaveBeenCalled();
  })

  it('creates Pictures on componentDidMount', async () => {
    const wrapper = shallow(<App />);
    wrapper.instance();
    wrapper.setState({ pictures })
    expect(wrapper.find(Pictures)).toBeDefined();
  })

  it('should call changeTab on click of Favorites', () => {
    wrapper = mount(<App />);
    const spy = jest.spyOn(wrapper.instance(), 'changeTab');
    wrapper.find(TabItem).at(1).simulate('click', 'using prototype');
    expect(spy).toHaveBeenCalled();
  })
  it('should call changeTab on click of All', () => {
    wrapper = mount(<App />);
    const spy = jest.spyOn(wrapper.instance(), 'changeTab');
    wrapper.find(TabItem).at(0).simulate('click', 'using prototype');
    expect(spy).toHaveBeenCalled();
  })

  it('should call changeTab when favorite is clicked', () => {
    wrapper = mount(<App />);
    wrapper.instance().setFavorite(1)
    expect(setFavoritePic).toHaveBeenCalled();
  })
})
