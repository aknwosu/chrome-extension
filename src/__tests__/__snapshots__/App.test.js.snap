// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="sc-bxivhb bIyjDb"
>
  <div
    className="view-tabs sc-ifAKCX dOlFsd"
  >
    <div
      className="sc-EHOje hGDLuQ"
      onClick={[Function]}
    >
      All
    </div>
    <div
      className="sc-EHOje eNhZsZ"
      onClick={[Function]}
    >
      Favorites
    </div>
  </div>
  <div>
    Loading...
  </div>
</div>
`;
